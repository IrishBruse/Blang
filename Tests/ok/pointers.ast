{
  "FunctionDeclarations": [
    {
      "Symbol": {
        "Name": "main",
        "Kind": 0
      },
      "Parameters": [],
      "Body": [
        {
          "_Type": "ExternalStatement",
          "Externals": [
            {
              "Name": "printf",
              "Kind": 0
            }
          ],
          "Range": {
            "Start": 13,
            "End": 26
          }
        },
        {
          "_Type": "AutoStatement",
          "Variables": [
            {
              "Name": "a",
              "Kind": 0
            },
            {
              "Name": "b",
              "Kind": 0
            },
            {
              "Name": "c",
              "Kind": 0
            }
          ],
          "Range": {
            "Start": 31,
            "End": 44
          }
        },
        {
          "_Type": "VariableDeclaration",
          "Symbol": {
            "Name": "a",
            "Kind": 0
          },
          "Value": {
            "_Type": "IntValue",
            "Value": 12,
            "Range": {
              "Start": 54,
              "End": 56
            }
          },
          "Range": {
            "Start": 50,
            "End": 51
          }
        },
        {
          "_Type": "FunctionCall",
          "Symbol": {
            "Name": "printf",
            "Kind": 0
          },
          "Parameters": [
            {
              "_Type": "StringValue",
              "Value": "a %d\\n",
              "Range": {
                "Start": 69,
                "End": 77
              }
            },
            {
              "_Type": "Variable",
              "Symbol": {
                "Name": "a",
                "Kind": 0
              },
              "Range": {
                "Start": 79,
                "End": 80
              }
            }
          ],
          "Range": {
            "Start": 62,
            "End": 68
          }
        },
        {
          "_Type": "VariableDeclaration",
          "Symbol": {
            "Name": "b",
            "Kind": 0
          },
          "Value": {
            "_Type": "AddressOfExpression",
            "Expr": {
              "_Type": "Variable",
              "Symbol": {
                "Name": "a",
                "Kind": 0
              },
              "Range": {
                "Start": 93,
                "End": 94
              }
            },
            "Range": {
              "Start": 93,
              "End": 94
            }
          },
          "Range": {
            "Start": 88,
            "End": 89
          }
        },
        {
          "_Type": "VariableDeclaration",
          "Symbol": {
            "Name": "c",
            "Kind": 0
          },
          "Value": {
            "_Type": "PointerDereferenceExpression",
            "Expr": {
              "_Type": "Variable",
              "Symbol": {
                "Name": "b",
                "Kind": 0
              },
              "Range": {
                "Start": 106,
                "End": 107
              }
            },
            "Range": {
              "Start": 106,
              "End": 107
            }
          },
          "Range": {
            "Start": 101,
            "End": 102
          }
        },
        {
          "_Type": "FunctionCall",
          "Symbol": {
            "Name": "printf",
            "Kind": 0
          },
          "Parameters": [
            {
              "_Type": "StringValue",
              "Value": "c %d\\n",
              "Range": {
                "Start": 120,
                "End": 128
              }
            },
            {
              "_Type": "Variable",
              "Symbol": {
                "Name": "c",
                "Kind": 0
              },
              "Range": {
                "Start": 130,
                "End": 131
              }
            }
          ],
          "Range": {
            "Start": 113,
            "End": 119
          }
        }
      ],
      "Range": {
        "Start": 0,
        "End": 135
      }
    }
  ],
  "GlobalVariables": [],
  "Range": {
    "Start": 0,
    "End": 135
  }
}