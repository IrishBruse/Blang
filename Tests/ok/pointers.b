main()
{
    extrn printf;
    auto a, b, c;

    a = 12;
    printf("a %d\n", a);

    b = &a;

    c = *b;
    printf("c %d\n", c);
}
