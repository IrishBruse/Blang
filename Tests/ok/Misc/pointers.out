a 12
c 12
